### ISSUE TITLE
vector iterator + offset out of range

### ISSUE THREAD

**Issue by jsprenkle** [Artelnics/opennn#80]
While trying the iris_plant example the program asserts during training.

File: C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\vector
Line: 159

Call stack:
1  std::_Vector_const_iterator<std::_Vector_val<std::_Simple_types<unsigned __int64>>>::operator+=            vector                  160   0x7ff63b091cb9 
2  std::_Advance<std::_Vector_const_iterator<std::_Vector_val<std::_Simple_types<unsigned __int64>>>,__int64> xutility                691   0x7ff63b08c142 
3  std::advance<std::_Vector_const_iterator<std::_Vector_val<std::_Simple_types<unsigned __int64>>>,__int64>  xutility                698   0x7ff63b08e900 
4  std::next<std::_Vector_const_iterator<std::_Vector_val<std::_Simple_types<unsigned __int64>>>>             xutility                779   0x7ff63b08f0f3 
5  OpenNN::Vector<unsigned __int64>::split                                                                    vector.h                11224 0x7ff63b0947cb 
6  OpenNN::Instances::get_training_batches                                                                    instances.cpp           1535  0x7ff63b0882cb 
7  OpenNN::CrossEntropyError::calculate_training_error                                                        cross_entropy_error.cpp 168   0x7ff63b21bb66 
8  OpenNN::QuasiNewtonMethod::perform_training                                                                quasi_newton_method.cpp 1838  0x7ff63b19131d 
9  main                                                                                                       main.cpp

**Comment by jsprenkle** [comment]
This code seems to work

template <class T>
Vector< Vector<T> > Vector<T>::split(const size_t& n) const
{
    // determine number of sub-vectors of size n
    const size_t batches_number = ( (this->size() - 1) / n ) + 1;

    // create array of vectors to store the sub-vectors
    Vector< Vector<T> > batches(batches_number);

    size_t remaining = this->size();

    for (size_t k = 0; k < batches_number; ++k)
    {
        // calculate size of this batch
        size_t batch_size = remaining > n ? n : remaining;

        // preallocate memory for the sub-vector
        batches[k].resize(batch_size);

        // copy elements from the input range to the sub-vector
        std::copy_n( begin(), batch_size, batches[k].begin() );

        remaining -= batch_size;
    }

    return batches;
}

**Comment by jsprenkle** [comment]
Or maybe not...

OpenNN. Iris Plant Application.
Training with quasi-Newton method...
Epoch 0;
Parameters norm: 7.56624
Training error: 209.687
Gradient norm: 0.200792
Training rate: 2.23721
Elapsed time: 01:21
Selection error: 66.5354
Epoch 5;
Parameters norm: 8.68111
Training error: 185.42
Gradient norm: 0.138022
Training rate: 0.106888
Elapsed time: 01:21
Selection error: 57.0846
Epoch 10;
Parameters norm: 8.8766
Training error: 185.405
Gradient norm: 0.123093
Training rate: 0.0522061
Elapsed time: 01:21
Selection error: 57.3809
Epoch 15;
Parameters norm: 9.04551
Training error: 185.399
Gradient norm: 0.112394
Training rate: 0.0656262
Elapsed time: 01:21
Selection error: 57.6405
Epoch 20;
Parameters norm: 9.20536
Training error: 185.397
Gradient norm: 0.103605
Training rate: 0.0656292
Elapsed time: 01:22
Selection error: 57.8861
Epoch 25;
Parameters norm: 9.39248
Training error: 185.395
Gradient norm: 0.0945431
Training rate: 0.982102
Elapsed time: 01:22
Selection error: 58.17
Epoch 30;
Parameters norm: 11.0682
Training error: 171.395
Gradient norm: 0.011559
Training rate: 22.0626
Elapsed time: 01:22
Selection error: 54.9928
Epoch 33: Minimum parameters increment norm reached.
Parameters increment norm: 0
Parameters norm: 37.3267
Training error: 156.893
Gradient norm: 1.00742e-006
Training rate: 0
Elapsed time: 01:22
Selection error: 45.3941
OpenNN Exception: ProbabilisticLayer class.
Vector<double> calculate_outputs(const Vector<double>&) const method.
Size must be equal to number of probabilistic neurons.

Press <RETURN> to close this window...

**Comment by alexey-zhurba** [comment]
Same issue here. Pretty cool project anyways!
Built with CMake 3.11.1 and Visual C++ 2017
<img width="1544" alt="error" src="https://user-images.githubusercontent.com/38475937/66151740-3ab42980-e618-11e9-871a-39e3c0c18af1.PNG">

**Comment by bubus112** [comment]
I got the same problem, have you managed to make it work?

**Comment by GuillermoArtelnics** [comment]
Hello ,

We have just uploaded a new version of `OpenNN`, where we have fixed the split method. 

The iris_plant example is now working correctly, I hope it was useful for you.

Thanks you for your help.
